src/main/java/jenkins/plugin/randomjobbuilder/GeneratorController.java

```diff
@@ -0,0 +1,365 @@
+import org.acegisecurity.context.SecurityContext;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Controls load generation for an input set of runs, by watching a registered set of LoadGenerators
+ * And using listeners to add/remove tasks as desired
+ * Registers run for lookup, so we know which ones to kill if we stop load suddenly...
+ *   or in which cases we need to start new jobs to maintain load level
+ *
+ *   And triggers new builds when needed
+ *
+ *
+ */
+@Restricted(NoExternalUse.class)
+@Extension
+public class GeneratorController extends RunListener<Run> {
+    private boolean autostart = false;
+
+    // FIXME the accounting for tasks is broken, sometimes it overcounts, sometimes it undercounts.
+    ConcurrentHashMap<String, LoadGenerator> registeredGenerators = new ConcurrentHashMap<String, LoadGenerator>();
+
+    /** Map {@link LoadGenerator#generatorId} to that LoadGenerator's queued item count */
+    ConcurrentHashMap<String, AtomicInteger> queueTaskCount = new ConcurrentHashMap<>();
+
+    /** Map {@link LoadGenerator#generatorId} to that LoadGenerator's {@link Run}s,
+     *   needed in order to cancel runs in progress */
+    ConcurrentHashMap<String, Collection<Run>> generatorToRuns = new ConcurrentHashMap<>();
+
+    /** Register the generator in {@link #registeredGenerators} or
+     *   replace existing generator with the same generator ID
+     *   Eventually we'll just use {@link ReconfigurableDescribable} for direct updates of generator lists.
+     *   @param generator Generator to register/update
+     */
+    public void registerOrUpdateGenerator(@Nonnull LoadGenerator generator) {
+        LoadGenerator previous = registeredGenerators.put(generator.getGeneratorId(), generator);
+        synchronized (generator) {
+            if (previous != null) {  // Copy some transitory state in, but honestly this is a hack
+                generator.copyStateFrom(previous);
+            }
+        }
+    }
+
+    /** Returns a snapshot of current registered load generators */
+    public List<LoadGenerator> getRegisteredGenerators() {
+        return new ArrayList<LoadGenerator>(registeredGenerators.values());
+    }
+
+    /**
+     * Unregister the generator and stop all jobs and tasks from it
+     * @param generator Generator to unregister/remove
+     */
+    public void unregisterAndStopGenerator(@Nonnull LoadGenerator generator) {
+        synchronized (generator) {
+            generator.stop();
+            this.stopAbruptly(generator);
+            registeredGenerators.remove(generator.getGeneratorId());
```

Comment:
seems suspicious... wouldn't the correct thing to do be remove the generator outside of the lock first?